src/Functions/FunctionBinaryArithmetic.h

```diff
@@ -1125,6 +1140,86 @@ class FunctionBinaryArithmetic : public IFunction
+    {
+        const auto * return_type_array = checkAndGetDataType<DataTypeArray>(result_type.get());
+
+        if (!return_type_array)
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Return type for function {} must be array.", getName());
+
+        if constexpr (is_multiply || is_division)
+            throw Exception(ErrorCodes::UNSUPPORTED_METHOD, "Cannot use multiplication or division on arrays");
+
+        auto num_args = arguments.size();
+        DataTypes data_types;
+
+        ColumnsWithTypeAndName new_arguments {num_args};
+        DataTypePtr result_array_type;
+
+        const auto * left_const = typeid_cast<const ColumnConst *>(arguments[0].column.get());
+        const auto * right_const = typeid_cast<const ColumnConst *>(arguments[1].column.get());
+
+        /// Unpacking arrays if both are constants.
+        if (left_const && right_const)
+        {
+            new_arguments[0] = {left_const->getDataColumnPtr(), arguments[0].type, arguments[0].name};
+            new_arguments[1] = {right_const->getDataColumnPtr(), arguments[1].type, arguments[1].name};
+            auto col = executeImpl(new_arguments, result_type, 1);
+            return ColumnConst::create(std::move(col), input_rows_count);
+        }
+
+        /// Unpacking arrays if at least one column is constant.
+        if (left_const || right_const)
+        {
+            new_arguments[0] = {arguments[0].column->convertToFullColumnIfConst(), arguments[0].type, arguments[0].name};
+            new_arguments[1] = {arguments[1].column->convertToFullColumnIfConst(), arguments[1].type, arguments[1].name};
+            return executeImpl(new_arguments, result_type, input_rows_count);
+        }
+
+        /// Unpacking non-const arrays and checking sizes of them.
+        if (*typeid_cast<const ColumnArray *>(arguments[0].column.get())->getOffsets().data() !=
+            *typeid_cast<const ColumnArray *>(arguments[1].column.get())->getOffsets().data())
+        {
+            throw Exception(ErrorCodes::SIZES_OF_COLUMNS_DOESNT_MATCH,
+            "Cannot apply operation for arguments of different sizes. Size of the first argument: {}, size of the second argument: {}",
+            *typeid_cast<const ColumnArray *>(arguments[0].column.get())->getOffsets().data(),
+            *typeid_cast<const ColumnArray *>(arguments[1].column.get())->getOffsets().data());
+        }
+
+        auto array_ptr = typeid_cast<const ColumnArray *>(arguments[0].column.get())->getData().getPtr();
+        result_array_type = typeid_cast<const DataTypeArray *>(arguments[0].type.get())->getNestedType();
+        new_arguments[0] = {array_ptr, result_array_type, arguments[0].name};
+
+        array_ptr = typeid_cast<const ColumnArray *>(arguments[1].column.get())->getData().getPtr();
+        result_array_type = typeid_cast<const DataTypeArray *>(arguments[1].type.get())->getNestedType();
+        new_arguments[1] = {array_ptr, result_array_type, arguments[1].name};
```

Comment:
```
        const auto * left_array_col = typeid_cast<const ColumnArray *>(arguments[0].column.get());
        const auto * right_array_col = typeid_cast<const ColumnArray *>(arguments[1].column.get());
        /// Unpacking non-const arrays and checking sizes of them.
        if (*left_array_col.data() != right_array_col ->getOffsets().data())
        {
            throw Exception(ErrorCodes::SIZES_OF_COLUMNS_DOESNT_MATCH,
            "Cannot apply operation for arguments of different sizes. Size of the first argument: {}, size of the second argument: {}",
            *left_array_col->getOffsets().data(),
            *right_array_col ->getOffsets().data());
        }

        const auto * left_array_type= typeid_cast<const DataTypeArray *>(arguments[0].type.get())->getNestedType();
        new_arguments[0] = {left_array_col , left_array_type, arguments[0].name};

        const auto * right_array_type= typeid_cast<const DataTypeArray *>(arguments[1].type.get())->getNestedType();
        new_arguments[1] = {right_array_col , right_array_type, arguments[1].name};
```

Have you thought about something like this? In my opinion it is more readable, but as it is just a personal opinion, feel free to ignore it.

P.S.: the code is not 100% complete, that is why it is not a suggestion.